File: go/ch28/benchmark.go
// go/ch28/benchmark.go
package main

import (
	"container/heap"
	"flag"
	"fmt"
	"os"
	"time"
)

const (
	gridN  = 50
	repeat = 400
	inf    = 1_000_000_000
)

type cell struct {
	d, r, c int
}

type pq []cell

func (h pq) Len() int            { return len(h) }
func (h pq) Less(i, j int) bool   { return h[i].d < h[j].d }
func (h pq) Swap(i, j int)        { h[i], h[j] = h[j], h[i] }
func (h *pq) Push(x interface{}) { *h = append(*h, x.(cell)) }
func (h *pq) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

func buildGrid(n int) [][]int {
	grid := make([][]int, n)
	for r := 0; r < n; r++ {
		grid[r] = make([]int, n)
		for c := 0; c < n; c++ {
			if (r+c)%7 == 0 {
				grid[r][c] = 1
			}
		}
	}
	grid[0][0] = 0
	grid[n-1][n-1] = 0
	return grid
}

func dijkstra(grid [][]int, sr, sc, tr, tc int) int {
	n := len(grid)
	dist := make([][]int, n)
	for i := range dist {
		dist[i] = make([]int, n)
		for j := range dist[i] {
			dist[i][j] = inf
		}
	}
	dist[sr][sc] = 0
	h := &pq{{0, sr, sc}}
	heap.Init(h)
	dirs := [][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
	for h.Len() > 0 {
		cur := heap.Pop(h).(cell)
		d, r, c := cur.d, cur.r, cur.c
		if d != dist[r][c] {
			continue
		}
		if r == tr && c == tc {
			return d
		}
		for _, dc := range dirs {
			nr, nc := r+dc[0], c+dc[1]
			if nr < 0 || nc < 0 || nr >= n || nc >= n {
				continue
			}
			if grid[nr][nc] == 1 {
				continue
			}
			nd := d + 1 + grid[nr][nc]
			if nd < dist[nr][nc] {
				dist[nr][nc] = nd
				heap.Push(h, cell{nd, nr, nc})
			}
		}
	}
	return inf
}

func runBenchmark() (float64, int) {
	grid := buildGrid(gridN)
	tr, tc := gridN-1, gridN-1
	checksum := 0
	start := time.Now()
	for rep := 0; rep < repeat; rep++ {
		for sr := 0; sr < gridN; sr += 5 {
			for sc := 0; sc < gridN; sc += 5 {
				if grid[sr][sc] == 1 {
					continue
				}
				checksum ^= dijkstra(grid, sr, sc, tr, tc)
			}
		}
	}
	elapsed := time.Since(start).Seconds()
	return elapsed, checksum
}

func main() {
	timingOnly := flag.Bool("timing-only", false, "print BENCHMARK_SEC/CHECKSUM to stderr and exit")
	flag.Parse()

	elapsed, checksum := runBenchmark()
	fmt.Fprintf(os.Stderr, "BENCHMARK_SEC=%.6f\n", elapsed)
	fmt.Fprintf(os.Stderr, "BENCHMARK_CHECKSUM=%d\n", checksum)
	if !*timingOnly {
		fmt.Fprintf(os.Stderr, "go: %.6fs checksum=%d\n", elapsed, checksum)
		fmt.Fprintf(os.Stderr, "hint: run python/ch28/benchmark.py for python/go speedup\n")
	}
}

File: python/ch28/benchmark.py
# python/ch28/benchmark.py
"""同一 Dijkstra ループの Python / Go ベンチマーク（第28章）."""
from __future__ import annotations

import heapq
import subprocess
import sys
import time
from pathlib import Path
from typing import List, Tuple

GRID_N = 50
REPEAT = 400
INF = 10**9


def build_grid(n: int) -> List[List[int]]:
    grid = [[0] * n for _ in range(n)]
    for r in range(n):
        for c in range(n):
            if (r + c) % 7 == 0:
                grid[r][c] = 1
    grid[0][0] = 0
    grid[n - 1][n - 1] = 0
    return grid


def dijkstra(grid: List[List[int]], sr: int, sc: int, tr: int, tc: int) -> int:
    n = len(grid)
    dist = [[INF] * n for _ in range(n)]
    dist[sr][sc] = 0
    pq: List[Tuple[int, int, int]] = [(0, sr, sc)]
    dirs = ((-1, 0), (1, 0), (0, -1), (0, 1))
    while pq:
        d, r, c = heapq.heappop(pq)
        if d != dist[r][c]:
            continue
        if r == tr and c == tc:
            return d
        for dr, dc in dirs:
            nr, nc = r + dr, c + dc
            if not (0 <= nr < n and 0 <= nc < n):
                continue
            if grid[nr][nc] == 1:
                continue
            nd = d + 1 + grid[nr][nc]
            if nd < dist[nr][nc]:
                dist[nr][nc] = nd
                heapq.heappush(pq, (nd, nr, nc))
    return INF


def run_benchmark() -> Tuple[float, int]:
    grid = build_grid(GRID_N)
    tr, tc = GRID_N - 1, GRID_N - 1
    checksum = 0
    start = time.perf_counter()
    for _ in range(REPEAT):
        for sr in range(0, GRID_N, 5):
            for sc in range(0, GRID_N, 5):
                if grid[sr][sc] == 1:
                    continue
                checksum ^= dijkstra(grid, sr, sc, tr, tc)
    elapsed = time.perf_counter() - start
    return elapsed, checksum


def repo_root() -> Path:
    return Path(__file__).resolve().parents[2]


def run_go_timing() -> Tuple[float, int]:
    go_file = repo_root() / "go" / "ch28" / "benchmark.go"
    proc = subprocess.run(
        ["go", "run", str(go_file), "--timing-only"],
        cwd=repo_root(),
        capture_output=True,
        text=True,
        check=False,
    )
    if proc.returncode != 0:
        print(proc.stderr, file=sys.stderr)
        raise RuntimeError(f"go benchmark failed: {proc.returncode}")
    elapsed = 0.0
    checksum = 0
    for line in proc.stderr.splitlines():
        if line.startswith("BENCHMARK_SEC="):
            elapsed = float(line.split("=", 1)[1])
        if line.startswith("BENCHMARK_CHECKSUM="):
            checksum = int(line.split("=", 1)[1])
    return elapsed, checksum


def main() -> None:
    py_sec, py_sum = run_benchmark()
    print(f"python: {py_sec:.6f}s  checksum={py_sum}", file=sys.stderr)

    try:
        go_sec, go_sum = run_go_timing()
    except (RuntimeError, FileNotFoundError) as exc:
        print(f"go compare skipped: {exc}", file=sys.stderr)
        return

    print(f"go:     {go_sec:.6f}s  checksum={go_sum}", file=sys.stderr)
    if go_sum != py_sum:
        print("warning: checksum mismatch (implementations differ?)", file=sys.stderr)
    if go_sec > 0:
        speedup = py_sec / go_sec
        print(f"speedup (python/go): {speedup:.2f}x", file=sys.stderr)


if __name__ == "__main__":
    main()
